File: structures/trie/python/trie.py
"""
    Very basic implementation
"""

class TrieNode():
    
    
    def __init__(self, char: None):
        self.char = char
        self.children = []

class Trie:
    def __init__(self):
        self.root = TrieNode('*')
    
    def add_word(self, word:str):
        node = self.root
        for char in word:
            found_in_child = False
            for child in node.children:
                if child.char == char:
                    found_in_child = True
                    node = child
                    break
            if not found_in_child:
                t = TrieNode(char)
                node.children.append(t)
                node = t
    
    def find(self, word:str):
        node = self.root
        counter = 0
        for char in word:
            found_in_child = False
            for child in node.children:
                if child.char == char:
                    counter += 1
                    found_in_child = True
                    node = child
                    break
            if not found_in_child:
                return False, counter
        return True, counter

if __name__ == '__main__':
    t = Trie()
    t.add_word("tree")
    t.add_word("troo")
    t.add_word("fix")
    assert t.find("tr") == (True, 2)
    assert t.find("treee") == (False, 4)
    assert t.find("troo") == (True, 4)
    assert len(t.root.children) == 2 

File: structures/trie/go/trie.go
package main

import (
	"fmt"
)

type ITrie interface {
	Add_word(word string)
	Find(word string) (bool, int)
}

type nodeTrie struct {
	char     string
	children []*nodeTrie
}

type Trie struct {
	root nodeTrie
}

func (t *Trie) Add_word(word string) {
	node := &t.root
	var foundInChildren bool
	for _, ch := range word {
		foundInChildren = false
		for _, child := range node.children {
			if child.char == string(ch) {
				foundInChildren = true
				node = child
				break
			}
		}
		if !foundInChildren {
			trie := nodeTrie{char: string(ch)}
			node.children = append(node.children, &trie)
			node = &trie
		}
	}
}

func (t *Trie) Find(word string) (bool, int) {
	node := t.root
	counter := 0
	var foundInChildren bool
	for _, ch := range word {
		foundInChildren = false
		for _, child := range node.children {
			if child.char == string(ch) {
				counter++
				foundInChildren = true
				node = *child

			}
		}
		if !foundInChildren {
			return false, counter
		}
	}
	return true, counter
}

func testTrie(trie ITrie) (bool, int) {
	trie.Add_word("frog")
	trie.Add_word("grod")
	return trie.Find("frof")
}

func main() {
	fmt.Println(testTrie(&Trie{}))
}
